src/Components/Components/src/RenderTree/EventHandlerDelegateWithClosuresComparer.cs

```diff
@@ -0,0 +1,130 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+using System.Collections.Concurrent;
+using System.Reflection;
+using System.Linq.Expressions;
+
+namespace Microsoft.AspNetCore.Components.RenderTree
+{
+    internal static class EventHandlerDelegateWithClosuresComparer
+    {
+        private static readonly ConcurrentDictionary<Type, Func<object, object, bool>> ClosureComparerCache = new();
+
+        internal static bool EventCallBackEquals(ref EventCallback left, ref EventCallback right)
+        {
+            if (Equals(left, right))
+            {
+                return true;
+            }
+
+            // Whenever the receivers are different and they are explicit we assume that
+            // the EventCallbacks are not equal.
+            if (left.Receiver != right.Receiver &&
+                (left.RequiresExplicitReceiver || right.RequiresExplicitReceiver))
+            {
+                return false;
+            }
+
+            return DelegateEquals(left.Delegate, right.Delegate);
+        }
+
+        internal static bool DelegateEquals(MulticastDelegate? left, MulticastDelegate? right)
+        {
+            if (Equals(left, right))
+            {
+                return true;
+            }
+
+            // If any of the delegates or their targets are null (static delegates) then normal
+            // equality is sufficient. No additional testing required.
+            if (left == null || right == null || left.Target == null || right.Target == null)
+            {
+                return false;
+            }
+
+            var oldTargetType = left.Target!.GetType();
+            var newTargetType = right.Target!.GetType();
+
+            // if the types are not the same, or the methods are not the same, these possible closures are
+            // not the same anyway.
+            if (oldTargetType != newTargetType || left.Method != right.Method)
+            {
+                return false;
+            }
+
+            // for speed we cache the comparison functions per closure type.
+            var comparison = ClosureComparerCache.GetOrAdd(newTargetType, GetClosureComparerForType);
+            return comparison(left.Target, right.Target);
+        }
+
+        private static Func<object, object, bool> GetClosureComparerForType([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicFields)] Type type)
+        {
+            // if this type is not compiler generated, it is not a closure, and the previous tests where
+            // enough to conclude that these are not equal. We just cache a delegate to a function that returns false.
+            if (!type.CustomAttributes.Any(x => x.AttributeType == typeof(CompilerGeneratedAttribute)))
+            {
+                return (objA, objB) => false;
```

Comment:
Can we put this in a static readonly field at the top of the class and just reference it here?

Reply:
We could, but the compiler is smart enough to create a single static instance of the lambda type and to keep returning that:
![image](https://user-images.githubusercontent.com/985974/114688985-8bc8d180-9d15-11eb-93b3-84c67de0b5e8.png)
It's pretty optimal already.

Reply:
make it a static delegate to avoid capturing state by mistake in the future

Reply:
Makes sense, will do.